File: flutter/android/src/main/kotlin/com/dkmads/ssp/flutter/DkmadsInstreamPlatformView.kt
package com.dkmads.ssp.flutter

import android.content.Context
import android.view.View
import android.widget.FrameLayout
import com.dkmads.ssp.DKMadsInstreamAdsLoader
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.StandardMessageCodec
import io.flutter.plugin.platform.PlatformView
import io.flutter.plugin.platform.PlatformViewFactory

class DkmadsInstreamViewFactory(
  private val messenger: BinaryMessenger,
  private val events: (Int, String, Map<String, Any?>) -> Unit,
) : PlatformViewFactory(StandardMessageCodec.INSTANCE) {
  override fun create(context: Context, viewId: Int, args: Any?): PlatformView {
    @Suppress("UNCHECKED_CAST")
    val params = args as? Map<String, Any?> ?: emptyMap()
    return DkmadsInstreamPlatformView(context, viewId, events, params)
  }
}

class DkmadsInstreamPlatformView(
  context: Context,
  private val viewId: Int,
  private val events: (Int, String, Map<String, Any?>) -> Unit,
  @Suppress("UNUSED_PARAMETER") creationParams: Map<String, Any?>,
) : PlatformView {
  private val container = FrameLayout(context)
  private val loader: DKMadsInstreamAdsLoader

  init {
    InstreamPlatformRegistry.put(viewId, this)
    loader = DKMadsInstreamAdsLoader(
      container,
      onPauseContent = { emit("pause_content") },
      onResumeContent = { emit("resume_content") },
    ).apply {
      listener = object : DKMadsInstreamAdsLoader.Listener {
        override fun onAdStarted(loader: DKMadsInstreamAdsLoader) {
          emit("ad_started")
        }

        override fun onAdFinished(loader: DKMadsInstreamAdsLoader) {
          emit("ad_finished")
        }

        override fun onAdFailed(loader: DKMadsInstreamAdsLoader, message: String) {
          emit("ad_failed", mapOf("message" to message))
        }
      }
    }
  }

  private fun emit(event: String, extra: Map<String, Any?> = emptyMap()) {
    events(viewId, event, extra)
  }

  fun requestAds(adUnitId: String, width: Int, height: Int, placementContext: String?) {
    loader.requestAds(
      adUnitId = adUnitId,
      contentPosition = placementContext,
      width = width,
      height = height,
    )
  }

  fun destroyLoader() {
    loader.destroy()
    InstreamPlatformRegistry.remove(viewId)
  }

  override fun getView(): View = container

  override fun dispose() {
    destroyLoader()
  }
}

object InstreamPlatformRegistry {
  private val views = mutableMapOf<Int, DkmadsInstreamPlatformView>()

  fun put(id: Int, view: DkmadsInstreamPlatformView) {
    views[id] = view
  }

  fun remove(id: Int) {
    views.remove(id)
  }

  fun get(id: Int): DkmadsInstreamPlatformView? = views[id]
}

File: flutter/ios/Classes/DkmadsInstreamPlatformView.swift
import Flutter
import UIKit

final class DkmadsInstreamPlatformView: NSObject, FlutterPlatformView {
  private let container = UIView()
  private let loader: DKMadsInstreamAdsLoader
  private let viewId: Int64
  private let onEvent: (Int64, String, [String: Any]) -> Void

  init(
    frame: CGRect,
    viewId: Int64,
    messenger: FlutterBinaryMessenger,
    onEvent: @escaping (Int64, String, [String: Any]) -> Void
  ) {
    self.viewId = viewId
    self.onEvent = onEvent
    container.frame = frame
    container.autoresizingMask = [.flexibleWidth, .flexibleHeight]
    container.isHidden = true
    loader = DKMadsInstreamAdsLoader(
      adContainer: container,
      onPauseContent: { onEvent(viewId, "pause_content", [:]) },
      onResumeContent: { onEvent(viewId, "resume_content", [:]) }
    )
    super.init()
    InstreamPlatformRegistry.shared.put(viewId: viewId, view: self)
    loader.delegate = self
  }

  func view() -> UIView { container }

  func requestAds(adUnitId: String, width: Int, height: Int, placementContext: String?) {
    loader.requestAds(
      adUnitID: adUnitId,
      contentPosition: placementContext,
      adSize: CGSize(width: width, height: height)
    )
  }

  func destroyLoader() {
    loader.destroy()
    InstreamPlatformRegistry.shared.remove(viewId: viewId)
  }

  deinit {
    destroyLoader()
  }
}

extension DkmadsInstreamPlatformView: DKMadsInstreamAdsLoaderDelegate {
  func instreamAdsLoaderDidStartAd(_ loader: DKMadsInstreamAdsLoader) {
    onEvent(viewId, "ad_started", [:])
  }

  func instreamAdsLoaderDidFinishAd(_ loader: DKMadsInstreamAdsLoader) {
    onEvent(viewId, "ad_finished", [:])
  }

  func instreamAdsLoader(_ loader: DKMadsInstreamAdsLoader, didFailWithError error: Error) {
    onEvent(viewId, "ad_failed", ["message": error.localizedDescription])
  }
}

final class DkmadsInstreamViewFactory: NSObject, FlutterPlatformViewFactory {
  private let messenger: FlutterBinaryMessenger
  private let onEvent: (Int64, String, [String: Any]) -> Void

  init(messenger: FlutterBinaryMessenger, onEvent: @escaping (Int64, String, [String: Any]) -> Void) {
    self.messenger = messenger
    self.onEvent = onEvent
    super.init()
  }

  func create(
    withFrame frame: CGRect,
    viewIdentifier viewId: Int64,
    arguments args: Any?
  ) -> FlutterPlatformView {
    DkmadsInstreamPlatformView(frame: frame, viewId: viewId, messenger: messenger, onEvent: onEvent)
  }

  func createArgsCodec() -> FlutterMessageCodec & NSObjectProtocol {
    FlutterStandardMessageCodec.sharedInstance()
  }
}

final class InstreamPlatformRegistry {
  static let shared = InstreamPlatformRegistry()
  private var views: [Int64: DkmadsInstreamPlatformView] = [:]

  func put(viewId: Int64, view: DkmadsInstreamPlatformView) {
    views[viewId] = view
  }

  func remove(viewId: Int64) {
    views.removeValue(forKey: viewId)
  }

  func get(viewId: Int64) -> DkmadsInstreamPlatformView? {
    views[viewId]
  }
}
